frame/benchmarking/README.md

```diff
@@ -1,3 +1,182 @@
+* Using the results of the benchmark to create a linear model of the function across its components.
+
+With this linear model, we are able to estimate ahead of time how long it would take to execute some
+logic, and thus make informed decisions without actually spending any significant resources.
+
+The benchmarking framework comes with the following components:
+
+* [A set of macros](./src/lib.rs) (`benchmarks!`, `add_benchmark!`, etc...) to make it easy to
+  write, test, and add runtime benchmarks.
+* [A set of linear regression analysis functions](./src/analysis.rs) for processing benchmark data.
+* [A CLI extension](../../utils/benchmarking-cli/) to make it easy to execute benchmarks on your
+  node.
+
+The end to end benchmarking pipeline is disabled by default when compiling a node. If you want to
+run benchmarks, you need to enable it by compiling with a Rust feature flag `runtime-benchmarks`.
+More details about this below.
+
+### Weight
+
+Substrate represents computational resources using a generic unit of measurement called "Weight". It
+defines 10^12 Weight as 1 second of computation on the physical machine used for benchmarking. This
+means that the weight of a function may change based on the specific hardware used to benchmark the
+runtime functions.
+
+By modeling the expected weight of each runtime function, the blockchain is able to calculate how
+many transactions or system level functions it will be able to execute within a certain period of
+time. Often, the limiting factor for a blockchain is the fixed block production time for the
+network.
+
+Within FRAME, each dispatchable function must have a `#[weight]` annotation with a function that can
+return the expected weight for the worst case scenario execution of that function given it's inputs.
+This benchmarking framework will result in a file that automatically generates those formulas for
+you which you can then use in your pallet.
+
+## Writing Benchmarks
+
+Writing a runtime benchmark is much like writing a unit test for your pallet. They need to be
+carefully crafted to execute a certain logical path in your code. In tests you want to check for
+various success and failure conditions, but with benchmarks you specifically want to test for the
+**most computationally heavy** path, a.k.a the "worst case scenario".
+
+This means that if there are certain storage items or runtime state that may effect the complexity
+of the function, for example triggering more iterations in a `for` loop, to get an accurate result,
+you must set up your benchmark to trigger this.
+
+It may be that there are multiple paths your function can go down, and it is not clear which one is
+most heavy. In this case, you should just create a benchmark for each scenario! You may find that
+there are paths in your code where complexity may become unbounded depending on user input. This may
+hint to you that you should implement sane boundaries for how a user can use your pallet. For
```

Comment:
```suggestion
there are paths in your code where complexity may become unbounded depending on user input. This may
be a hint that you should enforce sane boundaries for how a user can use your pallet. For
```